### ISSUE TITLE
TablePropertyClause caching must be a map

### ISSUE THREAD

**Issue by fggarcia** [outworkers/phantom#419]
Await.result(create
        .ifNotExists()
        .`with`(compression.eqs(SnappyCompressor))
        .`with`(gc_grace_seconds.eqs(CassandraConnector.config.gcGraceSeconds))
        .`with`(caching.eqs(KeysOnly))
        .future(), Duration(1, TimeUnit.MINUTES))

Invalid value for property 'caching'. It should be a map. akka.actor.ActorInitializationException: exception during creation

phantom-dsl version = 1.21.4

**Comment by alexflav23** [comment]
Hi @fggarcia,

Thank you for your bug report, we will fix this very shortly.

Regards.

**Comment by alexflav23** [comment]
Hi @fggarcia,

We are not able to replicate this locally, maybe this is required to be a map only when subproperties on caching are set. Do you have a concrete example and a stack trace that could be worked against?

Regards.

**Comment by alexflav23** [comment]
Hi @fggarcia,

Have a look at the tests [here](https://github.com/websudos/phantom/blob/develop/phantom-dsl/src/test/scala/com/websudos/phantom/builder/query/db/crud/CreateTest.scala), please let us know if you are doing anything differently.

Regards.

**Comment by fggarcia** [comment]
Class com.websudos.phantom.builder.query.Caching doesn't exist.

All my tables create with next code:

Await.result(create
        .ifNotExists()
        .`with`(compression.eqs(LZ4Compressor))
        .`with`(gc_grace_seconds.eqs(CassandraConnector.config.gcGraceSeconds))
        .`with`(compaction.eqs(LeveledCompactionStrategy))
        .`with`(caching eqs CacheStrategies.KeysOnly)
        .future(), Duration(1, TimeUnit.MINUTES))

caching requires 
def eqs(strategy: CacheProperty): TablePropertyClause = {
      new TablePropertyClause(QueryBuilder.Create.caching(strategy.qb.queryString))
    }

Only CacheProperty available

sealed class CacheProperty(val qb: CQLQuery)

object CacheStrategies {
  case object None extends CacheProperty(CQLQuery(CQLSyntax.CacheStrategies.None))
  case object KeysOnly extends CacheProperty(CQLQuery(CQLSyntax.CacheStrategies.KeysOnly))
}

**Comment by alexflav23** [comment]
Hi @fggarcia,

Try upgrading to `1.22.0` first.

**Comment by fggarcia** [comment]
Error persisst.

Await.result(create
        .ifNotExists()
        .`with`(compression.eqs(LZ4Compressor))
        .`with`(gc_grace_seconds.eqs(CassandraConnector.config.gcGraceSeconds))
        .`with`(compaction.eqs(LeveledCompactionStrategy))
        .`with`(caching eqs Caching.KeysOnly)
        .future(), Duration(1, TimeUnit.MINUTES))

2016-02-15T14:29:14.447Z (fede-lenovo-manjaro-v1.0.168-HOTEL-SNAPSHOT) ERROR [][][pool-20-thread-1] com.websudos.phantom - Invalid value for property 'caching'. It should be a map. ('127.0.0.1', 44750)
2016-02-15T14:29:14.447Z (fede-lenovo-manjaro-v1.0.168-HOTEL-SNAPSHOT) ERROR [][][pool-20-thread-2] com.websudos.phantom - Invalid value for property 'caching'. It should be a map. ('127.0.0.1', 44750)
2016-02-15T14:29:14.455Z (fede-lenovo-manjaro-v1.0.168-HOTEL-SNAPSHOT) ERROR [][][run-main-0] org.scalatra.servlet.ScalatraListener - Failed to initialize scalatra application at com.datastax.driver.core.exceptions.SyntaxError: Invalid value for property 'caching'. It should be a map.
	at com.datastax.driver.core.Responses$Error.asException(Responses.java:132) ~[cassandra-driver-core-3.0.0.jar:na]
	at com.datastax.driver.core.DefaultResultSetFuture.onSet(DefaultResultSetFuture.java:179) ~[cassandra-driver-core-3.0.0.jar:na]
	at com.datastax.driver.core.RequestHandler.setFinalResult(RequestHandler.java:184) ~[cassandra-driver-core-3.0.0.jar:na]
	at com.datastax.driver.core.RequestHandler.access$2500(RequestHandler.java:43) ~[cassandra-driver-core-3.0.0.jar:na]
	at com.datastax.driver.core.RequestHandler$SpeculativeExecution.setFinalResult(RequestHandler.java:798) ~[cassandra-driver-core-3.0.0.jar:na]
	at com.datastax.driver.core.RequestHandler$SpeculativeExecution.onSet(RequestHandler.java:617) ~[cassandra-dri ('127.0.0.1', 44750)

**Comment by alexflav23** [comment]
I think your problem is different. You should be using `with` followed by `and`.

Like this:

```scala
BasicTable.create
          .`with`(compaction eqs LeveledCompactionStrategy.sstable_size_in_mb(50.megabytes))
          .and(compression eqs LZ4Compressor.crc_check_chance(0.5))
```

**Comment by fggarcia** [comment]
Okey. I try:

create
        .ifNotExists()
        .`with`(compression.eqs(LZ4Compressor))
        .and(gc_grace_seconds.eqs(CassandraConnector.config.gcGraceSeconds))
        .and(compaction.eqs(LeveledCompactionStrategy))
        .and(caching eqs Caching.KeysOnly)
        .future(), Duration(1, TimeUnit.MINUTES))

But same result:

2016-02-15T14:39:09.181Z (fede-lenovo-manjaro-v1.0.168-HOTEL-SNAPSHOT) ERROR [][][pool-21-thread-1] com.websudos.phantom - Invalid value for property 'caching'. It should be a map. ('127.0.0.1', 49529)
2016-02-15T14:39:09.182Z (fede-lenovo-manjaro-v1.0.168-HOTEL-SNAPSHOT) ERROR [][][pool-21-thread-1] com.websudos.phantom - Invalid value for property 'caching'. It should be a map. ('127.0.0.1', 49529)
2016-02-15T14:39:09.193Z (fede-lenovo-manjaro-v1.0.168-HOTEL-SNAPSHOT) ERROR [][][run-main-0] org.scalatra.servlet.ScalatraListener - Failed to initialize scalatra application at com.datastax.driver.core.exceptions.SyntaxError: Invalid value for property 'caching'. It should be a map.
	at com.datastax.driver.core.Responses$Error.asException(Responses.java:132) ~[cassandra-driver-core-3.0.0.jar:na]
	at com.datastax.driver.core.DefaultResultSetFuture.onSet(DefaultResultSetFuture.java:179) ~[cassandra-driver-core-3.0.0.jar:na]
	at com.datastax.driver.core.RequestHandler.setFinalResult(RequestHandler.java:184) ~[cassandra-driver-core-3.0.0.jar:na]
	at com.datastax.driver.core.RequestHandler.access$2500(RequestHandler.java:43) ~[cassandra-driver-core-3.0.0.jar:na]
	at com.datastax.driver.core.RequestHandler$SpeculativeExecution.setFinalResult(RequestHandler.java:798) ~[cassandra-driver-core-3.0.0.jar:na]
	at com.datastax.driver.core.RequestHandler$SpeculativeExecution.onSet(RequestHandler.java:617) ~[cassandra-dri ('127.0.0.1', 49529)

**Comment by alexflav23** [comment]
Hi @fggarcia,

Which version of Cassandra are you using? You should also be able to do `query.queryString`, if you could println that and copy paste it here it would be immensely useful.

Regards.

**Comment by fggarcia** [comment]
I'm cassandra 3.0.3

queryString

CREATE TABLE IF NOT EXISTS memento_test.price_limits (key text, hotel text, value text, PRIMARY KEY (key, hotel)) WITH compression = { 'sstable_compression' : 'LZ4Compressor' } AND gc_grace_seconds = 259200 AND compaction = { 'class' : 'LeveledCompactionStrategy' } AND caching = 'keys_only';

**Comment by fggarcia** [comment]
It's should be caching = {'keys': 'ALL', 'rows_per_partition': 'NONE'}

**Comment by alexflav23** [comment]
Hi @fggarcia,

I think it's a Cassandra version variation, on my local Cassandra works perfectly fine. I will follow up with the Datastax team to figure out the cut off point and will fix this shortly.

Regards.

**Comment by alexflav23** [comment]
Hi @fggarcia,

It seems in more recent versions of Cassandra they changed the serialization, the work is now planned and there should be a new version out to fix this sometime this week. We've been kept in the dark about this particular update it seems.

Regards.

**Comment by fggarcia** [comment]
@alexflav23 Thanks!!

BTW: sorry for the question!! Is it possible to create materialized view with phantom???

**Comment by alexflav23** [comment]
Hi @fggarcia,

Yes, but we only offer it in phantom-pro, available from late April for a monthly subscription license. I will keep you up to date with that, it comes with a lot more(UDT, spark support etc).

Regards.

**Comment by alexflav23** [comment]
Hi @fggarcia,

This is now fixed in phantom `1.25.0`, going live today/tomorrow.
